### ISSUE TITLE
space optimization possible? struct{} takes up space when it is listed last in a struct

### ISSUE THREAD

**Issue by glycerine** [golang/go#17450]
I have a use case where it would be useful to have a field inside of a struct that gets changed in type from something to type struct{} -- and then no longer takes up space (I want to use golang structs like protobuf message definitions). 

When I checked my assumption however, I find a surprise: when the struct{} field is alone, it doesn't take up any space, as expected. But when the struct{} field is with another member, then it takes up 8 bytes. Ouch!  It seems like an opportunity for optimization has been missed.

```
package main

// run on : go version go1.7 darwin/amd64  (osx 10.11.6)

import (
	"fmt"
	"unsafe"
)

// A is size 0
type A struct {
	NotAMystery struct{} // takes up 0 bytes, as expected
}

// B is size 8
type B struct {
	Hithere int64 // takes up 8 bytes, as expected.
}

// C is size 16
type C struct {
	Hithere int64
	Mystery struct{} // takes up 8 bytes, HUH??
}

func main() {
	var a A
	var b B
	var c C
	fmt.Printf("size of A is '%v'\n", unsafe.Sizeof(a))
	fmt.Printf("size of B is '%v'\n", unsafe.Sizeof(b))
	fmt.Printf("size of C is '%v'\n", unsafe.Sizeof(c))
}

/*output

$ go run meas.go
size of A is '0'
size of B is '8'
size of C is '16'

*/
```

**Comment by glycerine** [comment]
So strange... if I move the struct{} field to the top, it again no longer takes up space!

```
// D is size 8                                                                                                              
type D struct {
        MoreMystery struct{} // takes up 0 bytes, now that it is at top.                                                    
        Hithere     int64
}

func main() {
        var a A
        var b B
        var c C
        var d D

        fmt.Printf("size of A is '%v'\n", unsafe.Sizeof(a))
        fmt.Printf("size of B is '%v'\n", unsafe.Sizeof(b))
        fmt.Printf("size of C is '%v'\n", unsafe.Sizeof(c))
        fmt.Printf("size of D is '%v'\n", unsafe.Sizeof(d))
}

/*output                                                                                                                    
                                                                                                                            
$ go run meas.go                                                                                                            
size of A is '0'                                                                                                            
size of B is '8'                                                                                                            
size of C is '16'                                                                                                           
size of D is '8'                                                                                                            
                                                                                                                            
*/
```

**Comment by glycerine** [comment]
The space loss seems happen only when the struct{} is the very last member of the struct.
 
```
// G is size 16                                                                                                             
type G struct {
        Starter int64
        Dep6    struct{} // takes up 0                                                                                      
        Hithere int64
}

// H is size 24                                                                                                             
type H struct {
        Starter int64
        Dep6    struct{} // takes up 0                                                                                      
        Hithere int64
        Dep7    struct{} // takes up 8                                                                                      
}

// I is size 24                                                                                                             
type I struct {
        Starter int64
        Hithere int64
        Dep6    struct{} // takes up 0                                                                                      
        Dep7    struct{} // takes up 8                                                                                      
}
```

**Comment by ianlancetaylor** [comment]
This is due to the fix for #9401.
